### DISCUSSION TITLE
insert or update on table * violates foreign key constraint

### DISCUSSION THREAD

**Discussion by leo-petrucci** [prisma/prisma#12263] [Q&A] [answered]
I'm having some trouble with the migration tool and I'm hoping someone can give me a hand.

I'm converting an existing schema/database to use Prisma, so far everything's gone great. I've created an initial migration, added a couple of new indexes.

But now I've hit a complete brick wall trying to add a `@relation` between two tables. 

These are my models:

```
// these have been simplified

model forums_archive_posts {
  archive_author_id      Int          @default(0)
  author                 core_members @relation(fields: [archive_author_id], references: [member_id])
}

model core_members {
  member_id                 Int                    @id
  forums_archive_posts      forums_archive_posts[] 
}
```

I've ran 

```
npx prisma migrate dev --name connected-posts-with-members
```

To create the migration. This reset my database, so I then restored a backup that didn't have the migration and ran

```
npx prisma migrate deploy
```

As I want to apply the local migration changes to my database. But I got this back:

```
Applying migration `20220311095752_connected_posts_with_members`
Error: P3018

A migration failed to apply. New migrations cannot be applied before the error is recovered from. Read more about how to resolve migration issues in a production database: https://pris.ly/d/migrate-resolve   

Migration name: 20220311095752_connected_posts_with_members

Database error code: 23503

Database error:
ERROR: insert or update on table "forums_archive_posts" violates foreign key constraint "forums_archive_posts_archive_author_id_fkey"
DETAIL: Key (archive_author_id)=(0) is not present in table "core_members".

DbError { severity: "ERROR", parsed_severity: Some(Error), code: SqlState("23503"), message: "insert or update on table \"forums_archive_posts\" violates foreign key constraint \"forums_archive_posts_archive_author_id_fkey\"", detail: Some("Key (archive_author_id)=(0) is not present in table \"core_members\"."), hint: None, position: None, where_: None, schema: Some("public"), table: Some("forums_archive_posts"), column: None, datatype: None, constraint: Some("forums_archive_posts_archive_author_id_fkey"), file: Some("ri_triggers.c"), line: Some(2463), routine: Some("ri_ReportViolation") }
```

I'm really confused, it seems to be saying that it's looking for `archive_author_id` in `core_members`, but I've never told it to do that. It should be using `member_id` instead.

Why isn't this working?

**Comment by leo-petrucci** [comment, marked as answer]
Right so this was probably just me not understanding databases/the error I was given.

```
DETAIL: Key (archive_author_id)=(0) is not present in table "core_members".
```
Literally means "There is no `core_members` row with `id = 0`.

So to fix it I went into my migration and added some SQL:

```
-- Update an existing user so I can use it as a default user
UPDATE core_members SET name = 'User' WHERE member_id = 1;

-- Change default value to 1, the id of the user above
ALTER TABLE "forums_archive_posts" ALTER COLUMN "archive_author_id" SET DEFAULT 1;
ALTER TABLE "forums_posts" ALTER COLUMN "author_id" SET DEFAULT 1;

-- Set existing posts to user above
UPDATE "forums_archive_posts" SET "archive_author_id" = 1 WHERE "archive_author_id" = 0;
UPDATE "forums_posts" SET "author_id" = 1 WHERE "author_id" = 0;
```

Then just 
```
npx prisma migrate deploy
```

And everything worked 👍

**Reply by Sujisujii** [reply]
How ti fix this issue in postgres

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
